harness/determined/pytorch/deepspeed/_deepspeed_trial.py

```diff
@@ -0,0 +1,1054 @@
+
+        If training without pipeline parallelism, users should implement this function by doing
+        the following things:
+
+        #. Get a batch from the dataloader_iter and pass it to the gpu.
+        #. Compute the loss in the forward pass.
+        #. Perform the backward pass.
+        #. Perform an optimizer step.
+        #. Return training metrics in a dictionary.
+
+        Here is a code example.
+
+        .. code-block:: python
+
+            # Assume one model_engine wrapped in ``__init__``.
+
+            batch = self.context.to_device(next(dataloader_iter))
+            loss = self.model_engine(batch)
+            self.model_engine.backward(loss)
+            self.model_engine.step()
+            return {"loss": loss}
+
+        If using gradient accumulation over multiple micro-batches, Determined will automatically
+        call ``train_batch``
+
+
+        With pipeline parallelism there is no need to manually get a batch from the dataloader_iter
+        and the forward, backward, optimizer steps are combined in the model engine's
+        ``train_batch`` method.
+
+        .. code-block:: python
+
+            # Assume one model_engine wrapped in ``__init__``.
+
+            loss = self.model_engine.train_batch(dataloader_iter)
+            return {"loss": loss}
+
+        Arguments:
+            dataloader_iter (Iterator[torch.utils.data.DataLoader]): iterator over a
+                torch DataLoader.
+            epoch_idx (integer): index of the current epoch among all the batches processed
+                per device (slot) since the start of training.
+            batch_idx (integer): index of the current batch among all the epoches processed
+                per device (slot) since the start of training.
+        Returns:
+            torch.Tensor or Dict[str, Any]:
+                training metrics to return.
+        """
+        pass
+
+    @abstractmethod
+    def build_training_data_loader(self) -> Optional[pytorch.DataLoader]:
+        """
+        Defines the data loader to use during training.
+
+        Must return an instance of :py:class:`determined.pytorch.DataLoader`.
+
+        If using data parallel training, the batch size should be per gpu batch size.
+        If using gradient aggregation, the dataloader should return batches with
+        `train_micro_batch_size_per_gpu` samples each.
+        """
+        pass
+
+    @abstractmethod
+    def build_validation_data_loader(self) -> Optional[pytorch.DataLoader]:
+        """
+        Defines the data loader to use during validation.
+
+        Must return an instance of :py:class:`determined.pytorch.DataLoader`.
+
+        If using data parallel training, the batch size should be per gpu batch size.
+        If using gradient aggregation, the dataloader should return batches with
+        `train_micro_batch_size_per_gpu` samples each.
+        """
+        pass
+
+    def build_callbacks(self) -> Dict[str, pytorch.PyTorchCallback]:
```

Comment:
Is it a good idea to reuse the PyTorchCallback class?

Reply:
I believe the same callback events apply for DeepSpeedTrial

Reply:
My guess is that if DeepSpeedTrial doesn't inherit for PyTorchTrial, that we'll eventually want a separate callback class too.  For now I guess it's ok, and we can probably make the change later in a backwards-compatible way.

Also, I [just landed](https://github.com/determined-ai/determined/pull/3643) more callbacks for PyTorchTrial.